ffs/minerselector/sr2/sr2.go

```diff
@@ -150,9 +159,27 @@ func getMinerQueryAsk(c *apistruct.FullNodeStruct, addrStr string) (uint64, erro
 		return 0, fmt.Errorf("getting miner %s info: %s", addr, err)
 	}
 
-	sask, err := c.ClientQueryAsk(ctx, *mi.PeerId, addr)
-	if err != nil {
-		return 0, fmt.Errorf("query asking: %s", err)
+	type chAskRes struct {
+		Error string
+		Ask   *storagemarket.StorageAsk
+	}
+	chAsk := make(chan chAskRes)
+	go func() {
+		sask, err := c.ClientQueryAsk(ctx, *mi.PeerId, addr)
+		if err != nil {
+			chAsk <- chAskRes{Error: err.Error()}
+			return
+		}
+		chAsk <- chAskRes{Ask: sask}
+	}()
+
+	select {
+	case <-time.After(time.Second * 10):
+		return 0, fmt.Errorf("query asking timed out")
+	case r := <-chAsk:
+		if r.Error != "" {
+			return 0, fmt.Errorf("query ask had controlled error: %s", r.Error)
+		}
+		return r.Ask.Price.Uint64(), nil
```

Comment:
An unfortunate change that I have to do to work around what I fixed [here](https://github.com/filecoin-project/go-fil-markets/pull/429). Since the API isn't respecting cancelations, I had to "abandon" the call manually.
Without this the selection takes a lot of time if some miners don't reply to the query-ask, since Lotus retries with exponential backoff and isn't respecting the cancellation.